### PR TITLE
Enabled compile-time validation of SQL template and its parameters

### PR DESCRIPTION
**Pull request by nakamura-to**
## Background
Komapper provides a means of executing SQL using [SQL templates](https://www.komapper.org/docs/reference/query/querydsl/template/). However, this approach has the following drawbacks:

- Syntax errors in SQL templates are not detected until runtime.
- There is a risk of missing parameter bindings.

To address these issues, we are introducing compile-time checks for SQL template syntax and parameter binding.

## Concept

We refer to a combination of an SQL template and its parameters as a command. There are three types of commands:

1. **One**: Retrieves a single record.
2. **Many**: Retrieves multiple records.
3. **Exec**: Executes operations such as insert, update, or delete.

Commands are defined using annotations, and the type of command is indicated by an interface.

During compilation, KSP is used to perform two main tasks:

- Check the SQL template syntax and ensure all parameters are bound.
- Generate extension functions for executing the command using QueryDsl.

## Example Command Definitions

Below are examples of how to define the `One`, `Many`, and `Exec` commands.

**One**
```kotlin
@KomapperCommand(
    """
    select * from author where id = /* id */0 limit 1
    """
)
class GetAuthor(val id: Int) : One<Author> {
    override fun TemplateSelectQueryBuilder.execute() = selectAsAuthor().single()
}
```

**Many**
```kotlin
@KomapperCommand(
    """
    select * from author order by name
    """
)
class ListAuthors : Many<Author> {
    override fun TemplateSelectQueryBuilder.execute() = selectAsAuthor()
}
```

**Exec**
```kotlin
@KomapperCommand(
    """
    insert into author (name, bio) values (/* name */'', /* bio */'')
    """
)
class CreateAuthor(val name: String, val bio: String) : Exec
```

For `One` and `Many`, an extension function `QueryDsl.fromCommand` is generated, while for `Exec`, the function `QueryDsl.executeCommand` is generated.

## Examples of Using the Extension Functions

```kotlin
// Select a single author
val getAuthor: Query<Address> = QueryDsl.fromCommand(GetAuthor(1))
```

```kotlin
// Select multiple authors
val listAuthors: Query<List<Author>> = QueryDsl.fromCommand(ListAuthors())
```

```kotlin
// Insert a new author
val command = CreateAuthor(
    name = "Brian Kernighan",
    bio = "Co-author of The C Programming Language and The Go Programming Language"
)
val createAuthor: Query<Long> = QueryDsl.executeCommand(command)
```

### CODE REVIEW

**Comment by dave08** [comment]
Little thought, maybe instead of overriding the function in One or Many, just make the user define a class deriving from a One with the constructor taking a lambda, then it could be much cleaner:

```kotlin
@KomapperCommand(
    """
    select * from author where id = /* id */0 limit 1
    """
)
class GetAuthor(val id: Int) : One { selectAsAuthor().single() }
```
is this possible?

**Comment by nakamura-to** [comment]
Thank you for the feedback. 
Your idea is interesting, but I believe it might be challenging to implement due to Kotlin’s syntax limitations.

If the `TemplateSelectQueryBuilder` notation is too long, we could consider creating a typealias to shorten it.

**Comment by dave08** [comment]
```kotlin
abstract class One<T>(val execute: TemplateSelectQueryBuilder.() -> Query<T>)
```

**Comment by nakamura-to** [comment]
I see!  
I was able to implement it like this:

```kotlin
@KomapperCommand(
    """
    select * from author where id = /* id */0 limit 1
    """
)
class GetAuthor(val id: Int) : One<Author>({ selectAsAuthor().single() })
```

This is an interesting idea. I’ll consider it further, and if there are no issues, I’d like to adopt it.

**Comment by dave08** [comment]
Another idea: 

```kotlin
fun interface One<T> { 
    fun TemplateSelectQueryBuilder.execute(): Query<T>
}

val authorGetter = One<Author> { selectAsAuthor().single() }

class GetAuthor(val id: Int) : One<Author> by authorGetter

fun main() {
	GetAuthor(1)
}
```
the advantage is if authorGetter is re-used in multiple queries, but it's at the cost of having to use delegates all the time. The previous method is also possible to declare a val for that part, but then the lambda needs the whole 

`val authorGetter: TemplateSelectQueryBuilder.() -> Query<T> = { ... }`, but then a typealias or `fun interface` for that part  would help there...

**Comment by nakamura-to** [comment]
Using a functional interface is a good idea. Inspired by the various ideas you provided, I’ve come up with a more flexible interface. How about the following approach?

Komapper could provide the following classes:

```kotlin
fun interface FetchOne<T> {
    fun TemplateSelectQueryBuilder.execute(): Query<T>
}

abstract class One<T> protected constructor(
    private val fetchOne: FetchOne<T>
) : FetchOne<T> by fetchOne
```

Applications can then use any of the following methods:

```kotlin
@KomapperCommand("""select * from author where id = /* id */0 limit 1""")
class GetAuthor(val id: Int) : One<Author>({ selectAsAuthor().single() })
```

```kotlin
// reusable function
val fetchAuthor = FetchOne { selectAsAuthor().single() }

@KomapperCommand("""select * from author where id = /* id */0 limit 1""")
class GetAuthor2(val id: Int) : One<Author>(fetchAuthor)
```

```kotlin
@KomapperCommand("""select * from author where id = /* id */0 limit 1""")
class GetAuthor3(val id: Int) : FetchOne<Author> {
    override fun TemplateSelectQueryBuilder.execute() = selectAsAuthor().single()
}
```

**Comment by nakamura-to** [comment]
I implemented the above idea in https://github.com/komapper/komapper/pull/1346/commits/65254afbc23d6a1f674f722c38eea498bd0438cb

File: komapper-annotation/src/main/kotlin/org/komapper/annotation/Annotations.kt [hunk 1]
```diff
@@ -274,6 +274,30 @@ annotation class KomapperProjectionDef(
...
+/**
+ * The annotated class is a command that encapsulates an SQL template, SQL parameters, and SQL execution into a single unit.
+ *
+ * @property sql the SQL template
+ * @property disableValidation if `true`, SQL validation at compile time will be disabled
+ */
+@Target(AnnotationTarget.CLASS)
+@Retention(AnnotationRetention.SOURCE)
+annotation class KomapperCommand(
+    val sql: String,
```
Comment (dave08) [inline komapper-annotation/src/main/kotlin/org/komapper/annotation/Annotations.kt:286]: Shouldn't there be `@Language("sql") ` on the sql property?

Comment (nakamura-to) [reply komapper-annotation/src/main/kotlin/org/komapper/annotation/Annotations.kt:286]: Good point. I will fix it.

**Comment by nakamura-to** [comment]
Now, returning records is supported with SQL templates:

```kotlin
@KomapperCommand(
    """
    insert into address 
        (address_id, street, version) 
    values 
        (/* id */0, /* street */'', /* version */0)
    returning
        address_id, street, version
    """,
)
class AddAddressThenReturn(val id: Int, val street: String, val version: Int) : ExecReturnOne<Address>({ selectAsAddress().single() })

val query: Query<Address> = QueryDsl.execute(AddAddressThenReturn(16, "STREET 16", 1))
```

Use `ExecReturnMany` instead of `ExecReturnOne` when executing operations like `UPDATE` or `DELETE` that return multiple rows.

**Comment by nakamura-to** [comment]
The `@KomapperPartial` annotation and the partial directive (`/*> ... */`) have been introduced. These allow you to embed fragments of SQL templates into other SQL templates.

```kotlin
@KomapperPartial(
    """
    /*%if pagination != null */
    limit /* pagination.limit */0 offset /*pagination.offset*/0
    /*%end*/
    """,
)
private const val paginationPartial = ""


@KomapperCommand(
    """
    select * from address order by address_id
    /*> paginationPartial */
    """,
)
class UsePartial(val pagination: Pagination?) : Many<Address>({ selectAsAddress() })

val query: Query<List<Address>> = QueryDsl.execute(UsePartial(Pagination(2, 3)))
```

The above query is converted into the following SQL:

```sql
select * from address order by address_id
limit ? offset ?
```

**Pull request merged**